src/proj_equirect.jl

```diff
@@ -242,41 +231,109 @@ end
+Base.firstindex(M::BlockDiagEquiRect) = 1
+Base.lastindex(M::BlockDiagEquiRect) = length(M)
+Base.iterate(M::BlockDiagEquiRect) = (Σ=parent(M) ; isempty(Σ) ? nothing : (Σ[:,:,1],1))
+Base.iterate(M::BlockDiagEquiRect, st) = st+1 > length(M) ? nothing : (parent(M)[:,:,st+1],  st+1)
+
+function Base.getindex(M::BlockDiagEquiRect, i::Int) 
+    1 <= i <= length(M) || throw(BoundsError(M, i))
+    return parent(M)[:,:,i]
 end
 
-function *(B::BlockDiagEquiRect{QUAzFourier}, f::EquiRectQUAzFourier)
-    # TODO: implement S2 multiplication
-    error("not implemented")
+function Base.setindex!(M::BlockDiagEquiRect, m::Matrix, i::Int)
+    1 <= i <= length(M) || throw(BoundsError(M, i))  
+    setindex!(parent(M)[:,:,i], m)
 end
 
-function adapt_structure(storage, L::BlockDiagEquiRect{B}) where {B}
-    BlockDiagEquiRect{B}(adapt(storage, L.blocks), adapt(storage, L.blocks_sqrt), adapt(storage, L.proj))
+# ## Mult (this one is different than M * f since )
+
+*(M::BlockDiagEquiRect{B}, f::EquiRectField) where {B<:Basis} = M * B(f)
+
+function *(M::BlockDiagEquiRect{AzFourier}, f::EquiRectAzFourier)
+    promote_metadata_strict(M.proj, f.proj) # ensure same projection
+    EquiRectAzFourier(@tullio(Bf[p,iₘ] := M.blocks[p,q,iₘ] * f.arr[q,iₘ]), f.metadata)
 end
 
-function simulate(rng::AbstractRNG, L::BlockDiagEquiRect{AzFourier,ProjEquiRect{T}}) where {T}
-    @unpack Ny, Nx, θspan = L.proj
-    z = EquiRectMap(randn(rng, T, Ny, Nx) .* sqrt.(sin.(range(θspan..., length=Ny))), L.proj)
-    sqrt(L) * z
+function *(M::BlockDiagEquiRect{QUAzFourier}, f::EquiRectQUAzFourier)
+    promote_metadata_strict(M.proj, f.proj) # ensure same projection
+    EquiRectQUAzFourier(@tullio(Bf[p,iₘ] := M.blocks[p,q,iₘ] * f.arr[q,iₘ]), f.metadata)
 end
 
+# ## map, e.g. map((Bℓ,fℓ)->sqrt(Bℓ)*fℓ, BlkDiag, field) -> field
+
+function Base.map(fun::Function, M::BlockDiagEquiRect{AzFourier}, f::F) where {F<:EquiRectAzFourier}
+    promote_metadata_strict(M.proj, f.proj) # ensure same projection
+    Mfarr = similar(f.arr)
+    for (Mfc, Mc, fc) in zip(eachcol(Mfarr), M, eachcol(f.arr))
+        Mfc = fun(Mc, fc)
+    end
+    EquiRectAzFourier(Mfarr, f.metadata)
+end 
+
+
+function Base.map(fun::Function, M::BlockDiagEquiRect{QUAzFourier}, f::F) where {F<:EquiRectQUAzFourier}
+    promote_metadata_strict(M.proj, f.proj) # ensure same projection
+    Mfarr = similar(f.arr)
+    for (Mfc, Mc, fc) in zip(eachcol(Mfarr), M, eachcol(f.arr))
+        Mfc = fun(Mc, fc)
+    end
+    EquiRectQUAzFourier(Mfarr, f.metadata)
+end 
+
+# ## I don't understand this yet ??
+# function adapt_structure(storage, L::BlockDiagEquiRect{B}) where {B}
+#     BlockDiagEquiRect{B}(adapt(storage, L.blocks), adapt(storage, L.blocks_sqrt), adapt(storage, L.proj))
+# end
+
+
+# # Question: does this need to match the pixel space size or the size of the
+# # matrix operator in the primal-dual stacked fourier domain
+# size(L::BlockDiagEquiRect) = (fill(L.proj.Nx * L.proj.Ny, 2)...,)
```

Comment:
For operators the `size` is only used in printing so it doesn't really matter, since multiplication is overriden directly rather than falling back to some generic `AbstractMatrix * AbstractVector` which would need `size`. That said, I think its good to try and keep `size` returning the actual size of the implicit matrix which this operator represents, which may be (and in this case, is) different than the size of the array which is actually stored in the operator. So basically, I think the above is fine, but its not a big deal.

Reply:
Yeah, its a bit tricky since the natural representation of these operators in my mind is more tensor-like rather than a straight up matrix. Its a lighter cognitive load for me if I leave it out ... so I'll go with that for now.

Reply:
Yea its definitely possible to think about these differently in different contexts, but at least for the purposes of the code, a field is a `<:AbstractVector` and a linear operator on that field is `<:AbstractMatrix` so if you're ever confused why something is the way is it or how it should be, its probably because its trying to adhere to this.